In [3]:
import numpy as np
import pandas as pd

In [4]:
orders = pd.read_csv("orders.csv")
orders = orders[orders['eval_set'] == 'prior']

In [5]:
orders['user_id'].nunique()

206209

In [6]:
orders_prior = pd.read_csv("order_products__prior.csv")

In [7]:
orders_prior

,order_id,product_id,add_to_cart_order,reordered
0,2,33120,1,1
1,2,28985,2,1
2,2,9327,3,0
3,2,45918,4,1
4,2,30035,5,0
...,...,...,...,...
32434484,3421083,39678,6,1
32434485,3421083,11352,7,0
32434486,3421083,4600,8,0
32434487,3421083,24852,9,1


In [8]:
products = pd.read_csv("products.csv")

In [9]:
departments = pd.read_csv("departments.csv")

In [10]:
orders.drop('eval_set',axis=1,inplace=True)

In [11]:
orders.sort_values(['user_id','order_number']).head(15)

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


In [12]:
products.head()

,product_id,product_name,aisle_id,department_id
0,1,Chocolate Sandwich Cookies,61,19
1,2,All-Seasons Salt,104,13
2,3,Robust Golden Unsweetened Oolong Tea,94,7
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1
4,5,Green Chile Anytime Sauce,5,13


In [13]:
departments.head()

,department_id,department
0,1,frozen
1,2,other
2,3,bakery
3,4,produce
4,5,alcohol


In [14]:
step_users = (orders.groupby('order_number')['user_id'].nunique().reset_index(name='users_reached').sort_values('order_number'))

In [15]:
step_users

,order_number,users_reached
0,1,206209
1,2,206209
2,3,206209
3,4,182223
4,5,162633
...,...,...
94,95,1592
95,96,1525
96,97,1471
97,98,1421


In [74]:
step_users['prev_users'] = step_users['users_reached'].shift(1)
step_users.head(10)

,order_number,users_reached,prev_users,retention_rate_prev,retention_rate_first_order,retention_percentage_prev,retention_percentage_first_order,dropoff_pct_prev,dropoff_pct_first_order
0,1,206209,NaN,NaN,1.000000,NaN,100.00,NaN,0.00
1,2,206209,206209.0,1.000000,1.000000,100.00,100.00,0.00,0.00
2,3,206209,206209.0,1.000000,1.000000,100.00,100.00,0.00,0.00
3,4,182223,206209.0,0.883681,0.883681,88.37,88.37,11.63,11.63
4,5,162633,182223.0,0.892494,0.788680,89.25,78.87,10.75,21.13
5,6,146468,162633.0,0.900604,0.710289,90.06,71.03,9.94,28.97
6,7,132618,146468.0,0.905440,0.643124,90.54,64.31,9.46,35.69
7,8,120918,132618.0,0.911777,0.586386,91.18,58.64,8.82,41.36
8,9,110728,120918.0,0.915728,0.536970,91.57,53.70,8.43,46.30
9,10,101696,110728.0,0.918431,0.493170,91.84,49.32,8.16,50.68


In [75]:
step_users.head(10)[['order_number','users_reached','prev_users']]

,order_number,users_reached,prev_users
0,1,206209,NaN
1,2,206209,206209.0
2,3,206209,206209.0
3,4,182223,206209.0
4,5,162633,182223.0
5,6,146468,162633.0
6,7,132618,146468.0
7,8,120918,132618.0
8,9,110728,120918.0
9,10,101696,110728.0


In [17]:
orders['user_id'].nunique()

206209

In [18]:
count_of_users_who_orderedatleastonce = float(step_users.sort_values('order_number').iloc[0]['users_reached'])
count_of_users_who_orderedatleastonce

206209.0

In [19]:
step_users['retention_rate_prev'] = step_users['users_reached']/step_users['prev_users']
step_users['retention_rate_first_order'] = step_users['users_reached']/count_of_users_who_orderedatleastonce

In [20]:
step_users

,order_number,users_reached,prev_users,retention_rate_prev,retention_rate_first_order
0,1,206209,NaN,NaN,1.000000
1,2,206209,206209.0,1.000000,1.000000
2,3,206209,206209.0,1.000000,1.000000
3,4,182223,206209.0,0.883681,0.883681
4,5,162633,182223.0,0.892494,0.788680
...,...,...,...,...,...
94,95,1592,1656.0,0.961353,0.007720
95,96,1525,1592.0,0.957915,0.007395
96,97,1471,1525.0,0.964590,0.007134
97,98,1421,1471.0,0.966010,0.006891


In [21]:
#step_users.loc[0, 'retention_rate_prev'] = 1
#step_users.loc[0, 'retention_rate_first_order'] = 1

In [22]:
step_users.head()

,order_number,users_reached,prev_users,retention_rate_prev,retention_rate_first_order
0,1,206209,NaN,NaN,1.000000
1,2,206209,206209.0,1.000000,1.000000
2,3,206209,206209.0,1.000000,1.000000
3,4,182223,206209.0,0.883681,0.883681
4,5,162633,182223.0,0.892494,0.788680


In [23]:
step_users['retention_percentage_prev'] = round(step_users['retention_rate_prev']*100,2)
step_users['retention_percentage_first_order'] = round(step_users['retention_rate_first_order']*100,2)

step_users['dropoff_pct_prev'] = round( 100 - step_users['retention_percentage_prev'],2)
step_users['dropoff_pct_first_order'] = round( 100 - step_users['retention_percentage_first_order'],2)



In [24]:
step_users.head(10)

,order_number,users_reached,prev_users,retention_rate_prev,retention_rate_first_order,retention_percentage_prev,retention_percentage_first_order,dropoff_pct_prev,dropoff_pct_first_order
0,1,206209,NaN,NaN,1.000000,NaN,100.00,NaN,0.00
1,2,206209,206209.0,1.000000,1.000000,100.00,100.00,0.00,0.00
2,3,206209,206209.0,1.000000,1.000000,100.00,100.00,0.00,0.00
3,4,182223,206209.0,0.883681,0.883681,88.37,88.37,11.63,11.63
4,5,162633,182223.0,0.892494,0.788680,89.25,78.87,10.75,21.13
5,6,146468,162633.0,0.900604,0.710289,90.06,71.03,9.94,28.97
6,7,132618,146468.0,0.905440,0.643124,90.54,64.31,9.46,35.69
7,8,120918,132618.0,0.911777,0.586386,91.18,58.64,8.82,41.36
8,9,110728,120918.0,0.915728,0.536970,91.57,53.70,8.43,46.30
9,10,101696,110728.0,0.918431,0.493170,91.84,49.32,8.16,50.68


In [25]:
step_users.sort_values('dropoff_pct_prev',ascending=False).head(2)

,order_number,users_reached,prev_users,retention_rate_prev,retention_rate_first_order,retention_percentage_prev,retention_percentage_first_order,dropoff_pct_prev,dropoff_pct_first_order
3,4,182223,206209.0,0.883681,0.883681,88.37,88.37,11.63,11.63
4,5,162633,182223.0,0.892494,0.788680,89.25,78.87,10.75,21.13


In [26]:
step_users.sort_values('dropoff_pct_prev',ascending=False).head(10)[['order_number','dropoff_pct_prev','dropoff_pct_first_order']]

,order_number,dropoff_pct_prev,dropoff_pct_first_order
3,4,11.63,11.63
4,5,10.75,21.13
5,6,9.94,28.97
6,7,9.46,35.69
7,8,8.82,41.36
8,9,8.43,46.30
9,10,8.16,50.68
10,11,7.68,54.47
11,12,7.41,57.84
12,13,7.17,60.87


The biggest single-step drop-off is at order 4→5 (21.31%). After order 5, 21.13% of users have dropped off entirely


 **Note:** This dataset only includes customers who placed at least 3 orders.
- Drop-off between orders 1→2 and 2→3 is not captured here.
- Real-world early churn is likely higher than what this funnel shows.

# Phase 2 — Segment Customers

In [27]:
orders.sort_values(['user_id','order_number']).head(10)

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


In [28]:
orders['avg_days_bw_orders'] = round(orders.groupby('user_id')['days_since_prior_order'].transform('median'),0)
orders

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order,avg_days_bw_orders
0,2539329,1,1,2,8,NaN,20.0
1,2398795,1,2,3,7,15.0,20.0
2,473747,1,3,3,12,21.0,20.0
3,2254736,1,4,4,7,29.0,20.0
4,431534,1,5,4,15,28.0,20.0
...,...,...,...,...,...,...,...
3421077,2558525,206209,9,4,15,22.0,20.0
3421078,2266710,206209,10,5,18,29.0,20.0
3421079,1854736,206209,11,4,10,30.0,20.0
3421080,626363,206209,12,1,12,18.0,20.0


In [29]:
orders.head(15)

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order,avg_days_bw_orders
0,2539329,1,1,2,8,NaN,20.0
1,2398795,1,2,3,7,15.0,20.0
2,473747,1,3,3,12,21.0,20.0
3,2254736,1,4,4,7,29.0,20.0
4,431534,1,5,4,15,28.0,20.0
5,3367565,1,6,2,7,19.0,20.0
6,550135,1,7,1,9,20.0,20.0
7,3108588,1,8,1,14,14.0,20.0
8,2295261,1,9,1,16,0.0,20.0
9,2550362,1,10,4,8,30.0,20.0


In [30]:
# Average days between orders for each user
order_cadence = orders[['user_id', 'avg_days_bw_orders']].drop_duplicates()

In [31]:
order_cadence.sort_values('user_id')

,user_id,avg_days_bw_orders
0,1,20.0
11,2,13.0
26,3,11.0
39,4,17.0
45,5,11.0
...,...,...
3420930,206205,20.0
3420934,206206,3.0
3421002,206207,8.0
3421019,206208,7.0


In [32]:
basket_size_perorder = orders.merge(orders_prior, how='inner', on='order_id').groupby(['user_id', 'order_id'])['product_id'].count().reset_index()

In [33]:
basket_size_perorder.rename(columns = {"product_id":"basket_size"}).sort_values('user_id').head(15)

,user_id,order_id,basket_size
7,1,2550362,9
6,1,2539329,5
5,1,2398795,6
4,1,2295261,6
3,1,2254736,5
2,1,550135,5
9,1,3367565,4
8,1,3108588,6
1,1,473747,5
0,1,431534,8


In [34]:
avg_basket_size = round(basket_size_perorder.groupby('user_id')['product_id'].median().reset_index().rename(columns={'product_id': 'avg_basket_size'}),0)
avg_basket_size.head(5)

,user_id,avg_basket_size
0,1,6.0
1,2,14.0
2,3,7.0
3,4,3.0
4,5,10.0


Basket size trend:

In [35]:
orders.sort_values(['user_id','order_number'])[['order_id','user_id','order_number']]

,order_id,user_id,order_number
0,2539329,1,1
1,2398795,1,2
2,473747,1,3
3,2254736,1,4
4,431534,1,5
...,...,...,...
3421077,2558525,206209,9
3421078,2266710,206209,10
3421079,1854736,206209,11
3421080,626363,206209,12


In [36]:
basket_size_perorder = orders_prior.groupby('order_id')['product_id'].nunique().reset_index().rename(columns = {"product_id":"basket_size"})

In [37]:
basket_size_perorder.head()

,order_id,basket_size
0,2,9
1,3,8
2,4,13
3,5,26
4,6,3


In [38]:
first2_orders = round(
    orders[orders['order_number'] <= 2]
    .merge(basket_size_perorder, how="left", on="order_id")
    .groupby('user_id')['basket_size'].median()
    .reset_index(), 0
)

In [39]:
first2_orders

,user_id,basket_size
0,1,6.0
1,2,10.0
2,3,10.0
3,4,3.0
4,5,10.0
...,...,...
206204,206205,12.0
206205,206206,3.0
206206,206207,17.0
206207,206208,11.0


In [40]:
first2_orders.rename(columns={"basket_size":"avg_basketsize_first2orders"})

,user_id,avg_basketsize_first2orders
0,1,6.0
1,2,10.0
2,3,10.0
3,4,3.0
4,5,10.0
...,...,...
206204,206205,12.0
206205,206206,3.0
206206,206207,17.0
206207,206208,11.0


In [41]:
last2_orders = round((basket_size_perorder
    .merge(orders, how='right', on='order_id')
    .groupby('user_id')
    .apply(lambda x: x.nlargest(2, 'order_number')['basket_size'].median())
    .reset_index()
    .rename(columns={0: 'avg_basket_size_last2_orders'})
),0)

/tmp/ipykernel_8812/3427312813.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: x.nlargest(2, 'order_number')['basket_size'].median())


In [42]:
last2_orders

,user_id,avg_basket_size_last2_orders
0,1,8.0
1,2,12.0
2,3,6.0
3,4,2.0
4,5,8.0
...,...,...
206204,206205,8.0
206205,206206,3.0
206206,206207,18.0
206207,206208,12.0


In [43]:
final = (order_cadence
    .merge(avg_basket_size, how='left', on='user_id')
    .merge(first2_orders, how='left', on='user_id')
    .merge(last2_orders, how='left', on='user_id')
)

In [44]:
final.shape

(206209, 5)

In [45]:
final.head()

,user_id,avg_days_bw_orders,avg_basket_size,basket_size,avg_basket_size_last2_orders
0,1,20.0,6.0,6.0,8.0
1,2,13.0,14.0,10.0,12.0
2,3,11.0,7.0,10.0,6.0
3,4,17.0,3.0,3.0,2.0
4,5,11.0,10.0,10.0,8.0


In [46]:
final.rename(columns={"basket_size":"avg_basket_size_first2_orders"},inplace=True)

In [47]:
final

,user_id,avg_days_bw_orders,avg_basket_size,avg_basket_size_first2_orders,avg_basket_size_last2_orders
0,1,20.0,6.0,6.0,8.0
1,2,13.0,14.0,10.0,12.0
2,3,11.0,7.0,10.0,6.0
3,4,17.0,3.0,3.0,2.0
4,5,11.0,10.0,10.0,8.0
...,...,...,...,...,...
206204,206205,20.0,8.0,12.0,8.0
206205,206206,3.0,4.0,3.0,3.0
206206,206207,8.0,13.0,17.0,18.0
206207,206208,7.0,13.0,11.0,12.0


In [48]:
final[final['avg_basket_size_first2_orders']>final['avg_basket_size_last2_orders']]

,user_id,avg_days_bw_orders,avg_basket_size,avg_basket_size_first2_orders,avg_basket_size_last2_orders
2,3,11.0,7.0,10.0,6.0
3,4,17.0,3.0,3.0,2.0
4,5,11.0,10.0,10.0,8.0
5,6,9.0,4.0,6.0,5.0
6,7,7.0,10.0,16.0,8.0
...,...,...,...,...,...
206187,206188,30.0,10.0,12.0,6.0
206189,206190,30.0,6.0,12.0,6.0
206196,206197,9.0,7.0,9.0,8.0
206197,206198,16.0,5.0,7.0,4.0


In [49]:
final[final['avg_basket_size_first2_orders']<final['avg_basket_size_last2_orders']]

,user_id,avg_days_bw_orders,avg_basket_size,avg_basket_size_first2_orders,avg_basket_size_last2_orders
0,1,20.0,6.0,6.0,8.0
1,2,13.0,14.0,10.0,12.0
8,9,18.0,30.0,20.0,23.0
9,10,18.0,30.0,26.0,30.0
10,11,22.0,13.0,13.0,16.0
...,...,...,...,...,...
206200,206201,9.0,12.0,14.0,17.0
206201,206202,13.0,9.0,11.0,12.0
206206,206207,8.0,13.0,17.0,18.0
206207,206208,7.0,13.0,11.0,12.0


In [50]:
final[final['avg_basket_size_first2_orders']==final['avg_basket_size_last2_orders']]

,user_id,avg_days_bw_orders,avg_basket_size,avg_basket_size_first2_orders,avg_basket_size_last2_orders
22,23,9.0,10.0,10.0,10.0
29,30,27.0,1.0,2.0,2.0
49,50,5.0,6.0,8.0,8.0
54,55,14.0,10.0,10.0,10.0
62,63,8.0,12.0,12.0,12.0
...,...,...,...,...,...
206173,206174,5.0,9.0,17.0,17.0
206182,206183,13.0,14.0,18.0,18.0
206202,206203,30.0,22.0,24.0,24.0
206203,206204,11.0,14.0,14.0,14.0


In [51]:
conditions = [
    final['avg_basket_size_first2_orders'] > final['avg_basket_size_last2_orders'],
    final['avg_basket_size_first2_orders'] == final['avg_basket_size_last2_orders'],
    final['avg_basket_size_first2_orders'] < final['avg_basket_size_last2_orders']
]

choices = ['Declining basket size', 'Same Basket size', 'Increasing basket size']

final['basket_size_trend'] = np.select(conditions, choices, default='Incomplete data')

In [52]:
final['basket_size_trend'].value_counts()

,count
basket_size_trend,
Increasing basket size,92616
Declining basket size,84083
Same Basket size,29510


In [53]:
final['basket_size_trend'].value_counts(normalize=True) * 100

,proportion
basket_size_trend,
Increasing basket size,44.913656
Declining basket size,40.775621
Same Basket size,14.310724


In [54]:
conditions = [
    (final['avg_days_bw_orders'] < 10) & (final['basket_size_trend'].isin(['Increasing basket size', 'Same Basket size'])),
    (final['avg_days_bw_orders'] > 20) | (final['basket_size_trend'] == 'Declining basket size')
]

choices = ['High-Value', 'At-Risk']

final['segment'] = np.select(conditions, choices, default='Neither')

In [55]:
final['segment'].value_counts(normalize=True)*100

,proportion
segment,
At-Risk,56.574640
High-Value,22.019892
Neither,21.405467


In [56]:
segment_summary = final.groupby('segment').agg(
    customer_count=('user_id', 'count'),
    avg_basket_size=('avg_basket_size', 'median'),
    avg_order_cadence=('avg_days_bw_orders', 'median')
).reset_index()

segment_summary

,segment,customer_count,avg_basket_size,avg_order_cadence
0,At-Risk,116662,8.0,19.0
1,High-Value,45407,8.0,7.0
2,Neither,44140,8.0,14.0


# Inference:

At-Risk customers buy just as much per order as High-Value customers in terms of basket size — the only difference is they come back half as often. This is a frequency problem, not a basket size problem.

In [57]:
# orders per month (30 days) --approx.
segment_summary['order_frequency'] = 30 / segment_summary['avg_order_cadence']

In [58]:
segment_summary

,segment,customer_count,avg_basket_size,avg_order_cadence,order_frequency
0,At-Risk,116662,8.0,19.0,1.578947
1,High-Value,45407,8.0,7.0,4.285714
2,Neither,44140,8.0,14.0,2.142857


In [59]:
segment_summary['revenue_proxy'] = segment_summary['avg_basket_size'] * segment_summary['order_frequency']
segment_summary

,segment,customer_count,avg_basket_size,avg_order_cadence,order_frequency,revenue_proxy
0,At-Risk,116662,8.0,19.0,1.578947,12.631579
1,High-Value,45407,8.0,7.0,4.285714,34.285714
2,Neither,44140,8.0,14.0,2.142857,17.142857


# Revenue Proxy

In [60]:
float(segment_summary[segment_summary['segment']=='High-Value']['revenue_proxy'].loc[1] / segment_summary[segment_summary['segment']=='At-Risk']['revenue_proxy'].loc[0] )

2.714285714285714

High-Value customers generate 2.7x the basket-frequency revenue proxy of At-Risk customers.

In [62]:
reordered_data = final[final['segment']=='At-Risk'].merge(orders,on='user_id',how ='left')[['user_id','order_number','order_id']].merge(orders_prior,on='order_id',how='left')[['user_id','order_number','order_id','product_id','reordered']].merge(products,on='product_id',how='left').merge(departments,on='department_id',how='left')

In [63]:
reordered_data.head()

,user_id,order_number,order_id,product_id,reordered,product_name,aisle_id,department_id,department
0,3,1,1374495,9387,0,Granny Smith Apples,24,4,produce
1,3,1,1374495,17668,0,Unsweetened Chocolate Almond Breeze Almond Milk,91,16,dairy eggs
2,3,1,1374495,15143,0,Blueberry Pint,24,4,produce
3,3,1,1374495,16797,0,Strawberries,24,4,produce
4,3,1,1374495,39190,0,Vanilla Unsweetened Almond Milk,91,16,dairy eggs


In [64]:
order1_depts = reordered_data[reordered_data['order_number'] == 1][['user_id', 'department']].drop_duplicates()

order3_depts = reordered_data[reordered_data['order_number'] == 3][['user_id', 'department']].drop_duplicates()

#We are comparing order 1 and order 3 here, because up until that point, all users on initially have been retained upto 3rd order. There is only drop off after the 3rd order

In [65]:
order1_depts.head()

,user_id,department
0,3,produce
1,3,dairy eggs
8,3,snacks
88,4,bakery
89,4,beverages


In [66]:
order1_depts['user_id'].nunique() #116662
order3_depts['user_id'].nunique() #116662

116662

In [67]:
# Users who bought from each dept in order 1
order1_counts = order1_depts.groupby('department')['user_id'].nunique().reset_index(name='users_per_department_order1')
order1_counts

,department,users_per_department_order1
0,alcohol,4181
1,babies,5596
2,bakery,35483
3,beverages,55562
4,breakfast,22056
5,bulk,1063
6,canned goods,29897
7,dairy eggs,81219
8,deli,32002
9,dry goods pasta,26408


In [68]:
both = order1_depts.merge(order3_depts, on=['user_id', 'department'], how='inner')
both

,user_id,department
0,3,produce
1,3,dairy eggs
2,3,snacks
3,4,bakery
4,4,beverages
...,...,...
351299,206203,pantry
351300,206203,frozen
351301,206205,dairy eggs
351302,206205,bakery


In [69]:
both_counts = both.groupby('department')['user_id'].nunique().reset_index(name='users_reordered')
both_counts

,department,users_reordered
0,alcohol,2088
1,babies,2477
2,bakery,17864
3,beverages,36282
4,breakfast,8549
5,bulk,243
6,canned goods,12323
7,dairy eggs,62699
8,deli,15289
9,dry goods pasta,10258


In [70]:
# Reorder rate on a department level
dept_reorder_rate = (order1_counts
    .merge(both_counts, on='department', how='left')
)
dept_reorder_rate['reorder_rate'] = dept_reorder_rate['users_reordered'] / dept_reorder_rate['users_per_department_order1']

# Top 5 leaking departments
dept_reorder_rate.sort_values('reorder_rate').head(5)

,department,users_per_department_order1,users_reordered,reorder_rate
15,other,1425,238,0.167018
12,international,10063,2273,0.225877
5,bulk,1063,243,0.228598
17,personal care,13372,3830,0.286419
14,missing,1072,326,0.304104


In [71]:
# Similarly to compare High Value customers
reordered_data_hv = final[final['segment']=='High-Value'].merge(orders,on='user_id',how='left')[['user_id','order_number','order_id']].merge(orders_prior,on='order_id',how='left')[['user_id','order_number','order_id','product_id','reordered']].merge(products,on='product_id',how='left').merge(departments,on='department_id',how='left')

order1_depts_hv = reordered_data_hv[reordered_data_hv['order_number'] == 1][['user_id', 'department']].drop_duplicates()
order3_depts_hv = reordered_data_hv[reordered_data_hv['order_number'] == 3][['user_id', 'department']].drop_duplicates()

order1_counts_hv = order1_depts_hv.groupby('department')['user_id'].nunique().reset_index(name='users_order1')
both_hv = order1_depts_hv.merge(order3_depts_hv, on=['user_id', 'department'], how='inner')
both_counts_hv = both_hv.groupby('department')['user_id'].nunique().reset_index(name='users_reordered')

dept_reorder_rate_hv = order1_counts_hv.merge(both_counts_hv, on='department', how='left')
dept_reorder_rate_hv['reorder_rate_hv'] = dept_reorder_rate_hv['users_reordered'] / dept_reorder_rate_hv['users_order1']

In [72]:
dept_reorder_rate_hv

,department,users_order1,users_reordered,reorder_rate_hv
0,alcohol,1500,980,0.653333
1,babies,2059,960,0.466246
2,bakery,10813,5613,0.519097
3,beverages,18530,12316,0.664652
4,breakfast,6446,2693,0.417778
5,bulk,376,99,0.263298
6,canned goods,7778,3322,0.427102
7,dairy eggs,28199,22290,0.790454
8,deli,9533,4700,0.493024
9,dry goods pasta,7031,2782,0.395676


In [73]:
comparison = dept_reorder_rate.merge(
    dept_reorder_rate_hv[['department', 'reorder_rate_hv']],
    on='department',
    how='left'
)
comparison['gap'] = comparison['reorder_rate_hv'] - comparison['reorder_rate']
comparison.sort_values('gap', ascending=False).head(5)

,department,users_per_department_order1,users_reordered,reorder_rate,reorder_rate_hv,gap
0,alcohol,4181,2088,0.499402,0.653333,0.153931
14,missing,1072,326,0.304104,0.360294,0.056190
5,bulk,1063,243,0.228598,0.263298,0.034700
19,produce,88378,73492,0.831564,0.863063,0.031499
4,breakfast,22056,8549,0.387604,0.417778,0.030174


# Inference:
----------------------------------------------------------------------------
- Personal care (13k users, 28.6%) — toothpaste, shampoo etc. are consumables people should be rebuying regularly. Low reorder rate here is a genuine signal that Instacart is losing that repeat purchase to somewhere else (Amazon, pharmacy). This is worth a targeted nudge
- Alcohol has the biggest gap (0.15) and a meaningful sample size (4,181 users)
----------------------------------------------------------------------------

- **Personal care** — lowest absolute reorder rate among At-Risk (28.6%)
- **Alcohol — largest** -  gap between At-Risk and High-Value (15 percentage points)